torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,55 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType typeToScalarType(const c10::TypePtr & type) {
+    if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    }
+    return scalarTypeFromJitType(type);
+  }
+
+  c10::ScalarType getPromotedTypeForArithmeticOp(Node *node) {
```

Comment:
This is really hard to follow and validate, especially without comments / tests

Reply:
I'm counting on test_number_math/test_tensor_number_math to ensure the dtypes we get back from these jit operations match those we get back in non-jit runs -- if there's a case you see missing I can add it.

I'll add some comments. Basically this is just trying to match the behavior in compute_common_type() in TensorIterator.cpp. Ideally we'd re-use more of the code, but it's a bit awkward since in jit we don't have actual tensors and so don't have as clear a representation of `wrapped_number`.